# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project 2: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: ** 

I think the kind of supervised learning to use here is classification, because I need to find which student(s) are more prone to abandon the school, and this kind of forecast is discrete, the variable isn't continuous.  For example, you can use a clustering method to classified the students in order to find which features of the students has the higher probability to abandon the school.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


The dataset used in this project is included as `student-data.csv`. This dataset has the following attributes:

- `school` : student's school (binary: "GP" or "MS")
- `sex` : student's sex (binary: "F" - female or "M" - male)
- `age` : student's age (numeric: from 15 to 22)
- `address` : student's home address type (binary: "U" - urban or "R" - rural)
- `famsize` : family size (binary: "LE3" - less or equal to 3 or "GT3" - greater than 3)
- `Pstatus` : parent's cohabitation status (binary: "T" - living together or "A" - apart)
- `Medu` : mother's education (numeric: 0 - none,  1 - primary education (4th grade), 2 - 5th to 9th grade, 3 - secondary education or 4 - higher education)
- `Fedu` : father's education (numeric: 0 - none,  1 - primary education (4th grade), 2 - 5th to 9th grade, 3 - secondary education or 4 - higher education)
- `Mjob` : mother's job (nominal: "teacher", "health" care related, civil "services" (e.g. administrative or police), "at_home" or "other")
- `Fjob` : father's job (nominal: "teacher", "health" care related, civil "services" (e.g. administrative or police), "at_home" or "other")
- `reason` : reason to choose this school (nominal: close to "home", school "reputation", "course" preference or "other")
- `guardian` : student's guardian (nominal: "mother", "father" or "other")
- `traveltime` : home to school travel time (numeric: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour)
- `studytime` : weekly study time (numeric: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours)
- `failures` : number of past class failures (numeric: n if 1<=n<3, else 4)
- `schoolsup` : extra educational support (binary: yes or no)
- `famsup` : family educational support (binary: yes or no)
- `paid` : extra paid classes within the course subject (Math or Portuguese) (binary: yes or no)
- `activities` : extra-curricular activities (binary: yes or no)
- `nursery` : attended nursery school (binary: yes or no)
- `higher` : wants to take higher education (binary: yes or no)
- `internet` : Internet access at home (binary: yes or no)
- `romantic` : with a romantic relationship (binary: yes or no)
- `famrel` : quality of family relationships (numeric: from 1 - very bad to 5 - excellent)
- `freetime` : free time after school (numeric: from 1 - very low to 5 - very high)
- `goout` : going out with friends (numeric: from 1 - very low to 5 - very high)
- `Dalc` : workday alcohol consumption (numeric: from 1 - very low to 5 - very high)
- `Walc` : weekend alcohol consumption (numeric: from 1 - very low to 5 - very high)
- `health` : current health status (numeric: from 1 - very bad to 5 - very good)
- `absences` : number of school absences (numeric: from 0 to 93)
- `passed` : did the student pass the final exam (binary: yes or no)

### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [2]:
# TODO: Calculate number of students
n_students = len(student_data.index)

# TODO: Calculate number of features
n_features = len(student_data.columns)-1  # The variable dependent must be subtracted 
                                          #in order to find the real features number.
# TODO: Calculate passing students
n_passed = len(student_data[(student_data["passed"])== "yes"])

# TODO: Calculate failing students
n_failed = n_students - n_passed

# TODO: Calculate graduation rate
grad_rate = float(n_passed) / n_students * 100

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [3]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [4]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [5]:
# TODO: Import any additional functionality you may need here
import random 
# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train
random.seed(123)
rows = random.sample(X_all.index, num_train)

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train = X_all.ix[rows]
X_test = X_all.drop(rows)
y_train = y_all.ix[rows]
y_test = y_all.drop(rows)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **
Classifier 1:  Support Vector Machine

- Describe one real-world application in industry where the model can be applied.

    Ans:/ Some applications can be found in text categorization, bioinformatics and image recognition. (Moguerza, Javier and Muñoz, Alberto. Support Vector Machines with Applications. arXiv:math/0612817v1 [math.ST] 28 Dec 2006)

- What are the strengths of the model; when does it perform well?
        
    Ans/: 
    - Effective in high dimensional spaces.
    - Still effective in cases where number of dimensions is greater than the number of samples.
    - Uses a subset of training points in the decision function (called support vectors), so it is also memory efficient.
    - Versatile: different Kernel functions can be specified for the decision function. Common kernels are provided, but it is also possible to specify custom kernels.
    source: http://scikit-learn.org/stable/modules/svm.html

- What are the weaknesses of the model; when does it perform poorly?
    
    Ans/:
    - If the number of features is much greater than the number of samples, the method is likely to give poor performances.
    - SVMs do not directly provide probability estimates, these are calculated using an expensive five-fold cross-validation (see Scores and probabilities, below).
    source: http://scikit-learn.org/stable/modules/svm.html

- What makes this model a good candidate for the problem, given what you know about the data?

    Ans/: Is a horse work in machine learning classification problems like this, for the size of the dimensions of the problem are not large enough to destabilize it (as is explained by the disadvantages of the model).


Classifier 2:  Stochastic Gradient Descent (SDG)

- Describe one real-world application in industry where the model can be applied.

    Ans:/ 
    -This algorithm is widely used in the field of neural networks (source: https://en.wikipedia.org/wiki/Stochastic_gradient_descent#Applications)
    

- What are the strengths of the model; when does it perform well? 

    Ans:/
    -Efficiency.
    -Ease of implementation (lots of opportunities for code tuning).
    source: http://scikit-learn.org/stable/modules/sgd.html

- What are the weaknesses of the model; when does it perform poorly?
    Ans/:
    -Stochastic Gradient Descent requires a number of hyperparameters such as the regularization parameter and the number of iterations.
    -Stochastic Gradient Descent is sensitive to feature scaling.
    source: http://scikit-learn.org/stable/modules/sgd.html
    
- What makes this model a good candidate for the problem, given what you know about the data?
    
    Ans:/ This method is very versatile and can be used in this kind of problems, . 

Classifier 3:  Logistic Regression

- Describe one real-world application in industry where the model can be applied.
    Ans/: There is a lot of examples of applications like credit risk analysis, marketing segmentation and forecasting  in medicine (source:  https://en.wikipedia.org/wiki/Logistic_regression#Fields_and_example_applications)
    
- What are the strengths of the model; when does it perform well? 

    -Allows properties of a linear regression model to be exploited
    -The logit itself can take values between - ∞ and + ∞
    -Probability remains constrained between 0 and 1
    -The logit can be directly related to odds of disease 
    source: https://onlinecourses.science.psu.edu/stat507/node/18

- What are the weaknesses of the model; when does it perform poorly?
    -requires large sample size to achieve stable results
    -may have multicollinearity
    -may over fit the data
    source: http://www.justanswer.com/1byy-statistical-analysis/6n506-advantages-disadvantages-logistic-regression.html

- What makes this model a good candidate for the problem, given what you know about the data?
    Ans:/ The logistic regression model is par excellence the first reference in classifications methods, is a probabilistic model that behaves very well and very easy to understand.


### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [6]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')



def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))
    print "\n"

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [7]:
# TODO: Import the three supervised learning models from sklearn
from sklearn.cross_validation import train_test_split
from sklearn.svm import SVC
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size = num_test, random_state = 123)

# TODO: Initialize the three models
clf_A = SVC(random_state=123)
clf_B = SGDClassifier(random_state=123)
clf_C = LogisticRegression(random_state=123)

# TODO: Set up the training set sizes
#rows_100 = random.sample(X_all.index, 100)
#rows_200 = random.sample(X_all.index, 200)

#X_train_100 = X_all.ix[rows_100]
#y_train_100 = y_all.ix[rows_100]

#X_train_200 = X_all.ix[rows_200]
#y_train_200 = y_all.ix[rows_200]

#X_train_300 = X_train
#y_train_300 = y_train


# TODO: Execute the 'train_predict' function for each classifier and each training set size

for size in [100, 200, 300]:
    globals()["predict_A_{}".format(size)] = train_predict(clf_A, X_train[:size], y_train[:size], X_test, y_test)
    globals()["predict_B_{}".format(size)] = train_predict(clf_B, X_train[:size], y_train[:size], X_test, y_test)
    globals()["predict_C_{}".format(size)] = train_predict(clf_C, X_train[:size], y_train[:size], X_test, y_test)

Training a SVC using a training set size of 100. . .
Trained model in 0.0210 seconds
Made predictions in 0.0060 seconds.
F1 score for training set: 0.8228.
Made predictions in 0.0050 seconds.
F1 score for test set: 0.8625.


Training a SGDClassifier using a training set size of 100. . .
Trained model in 0.0630 seconds
Made predictions in 0.0370 seconds.
F1 score for training set: 0.8205.
Made predictions in 0.0010 seconds.
F1 score for test set: 0.8718.


Training a LogisticRegression using a training set size of 100. . .
Trained model in 0.0320 seconds
Made predictions in 0.0010 seconds.
F1 score for training set: 0.8310.
Made predictions in 0.0010 seconds.
F1 score for test set: 0.8201.


Training a SVC using a training set size of 200. . .
Trained model in 0.0160 seconds
Made predictions in 0.0130 seconds.
F1 score for training set: 0.8182.
Made predictions in 0.0030 seconds.
F1 score for test set: 0.8734.


Training a SGDClassifier using a training set size of 200. . .
Trained mode

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1- SVC**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0210           |         0.0050         |       0.8228     |     0.8625      |
| 200               |        0.0160           |         0.0030         |       0.8182     |     0.8734      |
| 300               |        0.0160           |         0.0020         |       0.8515     |     0.8790      |

** Classifer 2- SGDClassifier**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0630           |         0.0010         |       0.8205     |     0.8718      |
| 200               |        0.0010           |         0.0000         |       0.7768     |     0.8415      |
| 300               |        0.0010           |         0.0000         |       0.7959     |     0.8625      |

** Classifer 3- LogisticRegression**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0030           |         0.0000         |        0.8310    |     0.8201      |
| 200               |        0.0050           |         0.0000         |        0.8188    |     0.8472      |
| 300               |        0.0060           |         0.0000         |        0.8326    |     0.8630      |


## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: **
I choose Support Vector Machine as best model by two reason: Because the F1 test score are the best for this problem. Maybe this model is not as fast as the logistic model, but in three settings has better F1 score, and I prefer the score rather than the efficiency over time, at least for a case like the one discussed in this exercise.  It would be good to also analyze in a sample of several million items which is better.  One would think that in very large samples the efficiency over time should not be significant, but right now I haven't arguments at the time to assert such a claim.

I think the running time changes depending on what applications are running on par with jupyther, or if all code runs at the same time, and usually the time scales are similar to those presented in this report, being always faster the logistic model , Followed by the stochastic model and being the most "slow" model SVM. This also makes I prefer, at least, in this situation the F1 score to time as a selection criterion

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **

Dear board of directors, in this problem I recommend using the Support Vector Machine model, this algorithm separates failed students from those who do not, taking data about previous students like their family group (parents education, parents job status, family size, etc.), school performance (weekly study time, number of past class failures, extra educational support, etc.) and others of a personal nature (with a romantic relationship, free time after school, etc.) with their respective statistics of success / failure; establishing boundaries between those who have succeeded and those who have failed.  This allows us to find patterns that are not so obvious and can be predicted by the SVC model.

In practice, when you are working in plane in order to find an equation which separate and classify two kind of data, when data is nonlinear is very difficult. To address such problems, you can apply the SVM which is an algorithm that capture nonlinearities well.  To do that, the method set a margin, between a boundary and the classified points, and this boundary will separate the data in the desired classification.  In order to find the best boundary, the algorithm must find the longest distance between the boundary and the adjacent points, this can be done by using the kernel "trick" or kernel functions.  It has been called a trick, because this functions use a very elegant mathematical solution, which at first glance seems like a trick.  In theory, to get the boundary you must project the points in a hyperplane and find the equation from each plane and optimize in order to find the best solution, and it is quite complicated.  

The beauty of kernel trick is in using the inner product of vectors (algebra linear operation) that can bypass the projection of each plane of hyperplane, and get the best solution boundary with the resulting vector (from the inner product) through an optimization process that maximizes the widest gap between classes and the boundary, and the output vectors are the support vectors.

The resulting boundary separate the classes in the desired classification, which will allow us to predict in the end and in a more sophisticated way, which students will fail and be able to intervene them before they drop out the studies as the board of directors expected to be achieved with this project.

The SVM algorithm was selected because it is simpler and it has a better predictive power than other alternatives evaluated like Stochastic Gradient Descend and Logistic Regression.  This model has a good fit with training data, and has a better capacity of forecast, given its F1 test score, and you could say that this model will be correct approximately 86% of the time.

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [8]:
start_2 = time()
# TODO: Import 'GridSearchCV' and 'make_scorer'
# from sklearn.model_selection import GridSearchCV -> Conda doesn´t update
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import f1_score, make_scorer

# TODO: Create the parameters list you wish to tune
parameters = {'kernel':('linear', 'rbf'), 'C':[1e-3, 1, 1e3],'random_state':[123],'gamma':[1e-1, 1, 1e1]}

# TODO: Initialize the classifier
clf = SVC()

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, param_grid = parameters, scoring = f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train,y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))
end_2 = time()
print "Grid Search Cross Validation  in {:.4f} seconds".format(end_2 - start_2)
print "F1 score for predicting all 'yes': {:.4f}".format(
        f1_score(y_true = ['yes']*n_passed + ['no']*n_failed, y_pred = ['yes']*n_students, pos_label='yes', 
                 average='binary'))

Made predictions in 0.0080 seconds.
Tuned model has a training F1 score of 0.9849.
Made predictions in 0.0020 seconds.
Tuned model has a testing F1 score of 0.8679.
Grid Search Cross Validation  in 76.1640 seconds
F1 score for predicting all 'yes': 0.8030


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0210           |         0.0050         |       0.8228     |     0.8625      |
| 200               |        0.0160           |         0.0030         |       0.8182     |     0.8734      |
| 300               |        0.0160           |         0.0020         |       0.8515     |     0.8790      |
| Grid search       |        0.0080           |         0.0020         |       0.9849     |     0.8679      |

The model obteined by Grid Search Cross Validation didn't obtain better results than those obtained by untuned model, only was better in execution training time, but this measure in this case is a little tricki, because the function only output the time of the selected model, doesn't display the total execution time of the grid search, the real execution time was 76.1640 seconds.  This last time would be worst or longer if the gamma parameter changes in a range for example of [1e-5,1e5], I did that and the execution time was approximately 2 hours and get the same F1 score.

This show me the selection and implementation of ML models goes beyond repeating a code and shows me that, also requires an intuition around the implementation of these codes to produce better results.

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.